jaxrs/WFLY-11943_Statistics_of_REST_endpoints.adoc

```diff
@@ -0,0 +1,112 @@
+:email:             rsearls@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org/browse
+
+== Overview
+
+Collect and report runtime statistics on REST endpoints running in RESTEasy.
+
+=== Project Scope
+
+Statistics reporting by Wildfly components (e.g. clustering, connector, ejb3, undertow)
+has existed from before Wildfly 7.0.0.Final.  The earliest introduction of this
+feature into Wildfly has not been found.  The feature defines XML attribute,
+`statistics-enabled` used in subsystem definitions to declare the default setting
+for data collection, and it is used in the CLI to turn on and off data collection,
+This feature also provides command `reset-statistics` to clear collected data and start a new.
+This RFE is adding component jaxrs to the set of reporting components using this feature.
+
+This RFE will not use or replace undertow's statistics data.  Endpoint data from
+within RESTeasy is reported.
+
+=== Not Project Scope
+
+This RFE is NOT related to Eclipse MicroProfile Metrics nor to the microprofile-metrics-smallrye.
+
+== Issue Metadata
+
+=== Issue
+
+* {issue-base-url}/WFLY-11943[WFLY-11943]
+* {issue-base-url}/EAP7-606[EAP7-606]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/WFLY-11751[WFLY-11751]
+* https://issues.jboss.org/browse/WFLY-12091[WFLY-12091]
+
+=== Dev Contacts
+
+* mailto:rsearls@redhat.com[Rebecca Searls]
+* mailto:rsigal@redhat.com[Ron Sigal]
+* mailto:asoldano@redhat.com[Alessio Soldano]
+* mailto:ema@redhat.com[Jim Ma]
+* mailto:weli@redhat.com[Wei Nan Li]
+
+=== QE Contacts
+
+* mailto:mkopecky@redhat.com[Marek Kopecky]
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+[X] Engineering
+
+[ ] QE
+
+=== Affected Projects or Components
+
+* WildFly
+* RESTEasy
+
+=== Other Interested Projects
+
+* HAL
+* CLI
+
+== Requirements
+
+* Collect endpoint statistics when instructed by Wildfly.
+* Report accumulated statistics when requested by Wildfly.
+* Reset endpoint statistics when instructed by Wildfly.
+* Stop data collection and clear accumumlated data when instructed by Wildfly.
+* Subsystem attribute `statistics-enabled` will be used to start and stop statistics collection.  Default setting is false.  The default value can be declared to `true` in the Wildfly configuration file, jaxws subsystem element.
+* Subsystem operation `read-resource` will display the statistics for each endpoint.
+* Subsystem operation `reset` will reset collected statistics to zero.
+* Application restart is NOT required when enabling, disabling and resetting statistics collection.
+* Statistics are NOT save on application restart.
+
+=== Hard Requirements
+* Statistics provided
+** `invocation-count` - Total count of calls to endpoint
+** `failed-invocation-count` - Number of endpoint calls that failed
```

Comment:
Can you please elaborate what the fail means here? Is it exception from end-point only? What if user does something like:
```java
if (error) {
    return Response.status(500).build();
}
```

Reply:
>  why MP Metrics should not be used
>  why the solution without MP Metrics is better

Upon starting this work I reviewed the MicroProfile Metrics specification and I 
investigated how other wfly components were currently using it to provide "vendor"
metrics.  I did not find other wfly components providing "vendor" metrics with
this spec, and I am not aware of plans by these components to move in that direction
yet.

My evaluation of the MicroProfile Metrics API is that it will add a level of complexity, 
larger footprint and most likely an undesirable performance impact to resteasy.  
In addition the spec and implementations like smallrye-metrics seems to still be evolving.  
It is for these reasons I chose not to use any MicroProfile Metrics APIs at this 
time and provided a simpler more efficient implementation that leverages the existing
resteasy classes and is in line with what other wfly components have implemented.


> Can you please elaborate what the fail means here? Is it exception from end-point 
> only? What if user does something like:
In this case fail means exception thrown.  A response object returning any
status is not considered a failure in this case.

Reply:
@rsearls Thank you for the clarification. Can you please add these information (from your previous comment) to the analysis? After that (and based on mail conversation with @bstansberry from 2019_07_15) I will approve this PR with analysis and we can move this RFE forward ...

Reply:
Done.